allennlp/data/dataset_readers/coreference_resolution/preco.py

```diff
@@ -0,0 +1,89 @@
+import json
+import logging
+from typing import Dict, List, Optional, Tuple
+
+from overrides import overrides
+
+from allennlp.common.file_utils import cached_path
+from allennlp.data.dataset_readers.coreference_resolution.conll import ConllCorefReader
+from allennlp.data.dataset_readers.dataset_reader import DatasetReader
+from allennlp.data.instance import Instance
+from allennlp.data.tokenizers import PretrainedTransformerTokenizer
+from allennlp.data.token_indexers import TokenIndexer
+
+logger = logging.getLogger(__name__)
+
+
+@DatasetReader.register("preco")
+class PrecoReader(DatasetReader):
+    """
+    Reads a single JSON-lines file. Each line contains a "sentences" key for a list of sentences
+    and a "mention_clusters" key for the clusters.
+
+    Returns a `Dataset` where the `Instances` have four fields : `text`, a `TextField`
+    containing the full document text, `spans`, a `ListField[SpanField]` of inclusive start and
+    end indices for span candidates, and `metadata`, a `MetadataField` that stores the instance's
+    original text. For data with gold cluster labels, we also include the original `clusters`
+    (a list of list of index pairs) and a `SequenceLabelField` of cluster ids for every span
+    candidate.
+
+    # Parameters
+
+    max_span_width : `int`, required.
+        The maximum width of candidate spans to consider.
+    token_indexers : `Dict[str, TokenIndexer]`, optional
+        This is used to index the words in the document.  See :class:`TokenIndexer`.
+        Default is `{"tokens": SingleIdTokenIndexer()}`.
+    wordpiece_modeling_tokenizer: `PretrainedTransformerTokenizer`, optional (default = None)
+        If not None, this dataset reader does subword tokenization using the supplied tokenizer
+        and distribute the labels to the resulting wordpieces. All the modeling will be based on
+        wordpieces. If this is set to `False` (default), the user is expected to use
+        `PretrainedTransformerMismatchedIndexer` and `PretrainedTransformerMismatchedEmbedder`,
+        and the modeling will be on the word-level.
+    """
+
+    def __init__(
+        self,
+        max_span_width: int,
+        token_indexers: Dict[str, TokenIndexer] = None,
+        wordpiece_modeling_tokenizer: Optional[PretrainedTransformerTokenizer] = None,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs)
+        self._conll_reader = ConllCorefReader(
+            max_span_width, token_indexers, wordpiece_modeling_tokenizer, **kwargs
```

Comment:
rather than doing this funny composition, it might be nice to pull out the `text_to_instance` method of the `ConllCorefReader` into a `coreference_resolution/util.py` file, [like we have done for reading comprehension](https://github.com/allenai/allennlp-reading-comprehension/blob/master/allennlp_rc/dataset_readers/util.py#L163).

This might be good, because I know there is a new dataset that lots of people are evaluating on - [GAP](https://github.com/google-research-datasets/gap-coreference). If we wanted to add that as well, it would make it really easy.